src/credentials/FabricTable.cpp

```diff
@@ -1248,97 +1378,546 @@ CHIP_ERROR FabricTable::AllocatePendingOperationalKey(Optional<FabricIndex> fabr
+                                                  FabricIndex * outNewFabricIndex)
+{
+    VerifyOrReturnError(mOpCertStore != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(outNewFabricIndex != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    static_assert(kMaxValidFabricIndex <= UINT8_MAX, "Cannot create more fabrics than UINT8_MAX");
+
+    // We should already have a pending root when we get here
+    VerifyOrReturnError(mStateFlags.Has(StateFlags::kIsTrustedRootPending), CHIP_ERROR_INCORRECT_STATE);
+    // We should not have pending update when we get here
+    VerifyOrReturnError(!mStateFlags.Has(StateFlags::kIsUpdatePending), CHIP_ERROR_INCORRECT_STATE);
+
+    EnsureNextAvailableFabricIndexUpdated();
+    FabricIndex fabricIndexToUse = kUndefinedFabricIndex;
+    if (mNextAvailableFabricIndex.HasValue())
+    {
+        // Mark we have some pending data for a given fabric.
+        fabricIndexToUse = mNextAvailableFabricIndex.Value();
+    }
+    else
+    {
+        // Fabric addition, but adding fabric would fail on table full: let's not allocate a fabric
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    if (existingOpKey == nullptr)
+    {
+        // If existing operational key not provided, we need to have a keystore present.
+        // It should already have an operational key pending.
+        VerifyOrReturnError(mOperationalKeystore != nullptr, CHIP_ERROR_KEY_NOT_FOUND);
+        // Make sure we have an operational key, pending or not
+        VerifyOrReturnError(mOperationalKeystore->HasOpKeypairForFabric(fabricIndexToUse) ||
+                                mOperationalKeystore->HasPendingOpKeypair(),
+                            CHIP_ERROR_KEY_NOT_FOUND);
+    }
+
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndexToUse), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    // Check for new fabric colliding with an existing fabric
+    if (!mStateFlags.Has(StateFlags::kAreCollidingFabricsIgnored))
+    {
+        FabricIndex collidingFabricIndex = kUndefinedFabricIndex;
+        ReturnErrorOnFailure(FindExistingFabricByNocChaining(fabricIndexToUse, noc, collidingFabricIndex));
+        ReturnErrorCodeIf(collidingFabricIndex != kUndefinedFabricIndex, CHIP_ERROR_FABRIC_EXISTS);
+    }
+
+    // We don't have a collision, handle the temp insert of NOC/ICAC
+    ReturnErrorOnFailure(mOpCertStore->AddNewOpCertsForFabric(fabricIndexToUse, noc, icac));
+    VerifyOrReturnError(SetPendingDataFabricIndex(fabricIndexToUse), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR err =
+        AddOrUpdateInner(kUndefinedFabricIndex, existingOpKey, isExistingOpKeyExternallyOwned, vendorId, outNewFabricIndex);
+    if (err != CHIP_NO_ERROR)
+    {
+        // Revert partial state added on error
+        mOpCertStore->RevertPendingOpCertsExceptRoot();
+        return err;
+    }
+
+    if (fabricIndexToUse != *outNewFabricIndex)
```

Comment:
Can we not just change the fabric index out parameter for `AddOrUpdateInner` into an input parameter and have it check this on behalf of the caller?

The need to predetermine the fabric index and then verify that `AddOrUpdateInner` also used this index seems to slightly complicate things.

Reply:
Will do as follow-up